strawberry_django/auth/utils.py

```diff
@@ -0,0 +1,65 @@
+from typing import Literal, Optional, overload
+
+from asgiref.sync import sync_to_async
+from strawberry.types import Info
+
+from strawberry_django.utils.typing import UserType
+
+
+@overload
+def get_current_user(info: Info, *, strict: Literal[True]) -> UserType: ...
+
+
+@overload
+def get_current_user(info: Info, *, strict: bool = False) -> Optional[UserType]: ...
+
+
+def get_current_user(info: Info, *, strict: bool = False) -> Optional[UserType]:
+    """Get and return the current user based on various scenarios."""
+    try:
+        user = info.context.request.user
+    except AttributeError:
+        try:
+            # queries/mutations in ASGI move the user into consumer scope
+            user = info.context.get("request").consumer.scope["user"]
+        except AttributeError:
+            # websockets / subscriptions move scope inside of the request
+            try:
+                user = info.context.get("request").scope.get("user")
+            except AttributeError:
+                # When running tests, we must manually add a context.
+                # For now let's grab it here:
+                is_test_mode = info.context.get("is_test_mode")
+                if is_test_mode:
```

Comment:
I'm not a huge fan of this. Usually we would want to mock this or solve it in another way

What issue were you trying to fix with this code?

Reply:
Oh, I just saw your comment in the PR. I'll reply there

Reply:
I like the current approach much better indeed.  My addition was reverted, and the docs which were missing the `with` statement are fixed for whoever needs it after me.